std/numeric.d

```diff
@@ -3329,6 +3329,117 @@ C swapRealImag(C)(C input)
     return C(input.im, input.re);
 }
 
+/**
+Finds the mean (colloquially known as the average) of a range.
+
+For built-in numerical types, accurate Knuth & Welford mean calculation
+is used. The default result type is `real`, which is more accurate,
+but far slower than `double` on many architectures.
+
+For user defined types, element by element summation is used. Additionally
+an extra parameter `seed` is needed in order to correctly seed the summation
+with the equivalent to `0`.
+
+The first overload of this function will return `T.init` if the range
+is empty. However, the second overload will return `seed` on empty ranges.
+
+This function is $(BIGOH r.length).
+
+Params:
+    r = An $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    seed = For user defined types. Should be equivalent to `0`.
+
+Returns:
+    The mean of `r` when `r` is non-empty.
+*/
+T mean(T = real, R)(R r)
+    if (isInputRange!R &&
+        isNumeric!(ElementType!R) &&
+        !isInfinite!R)
+{
+    if (r.empty)
+        return T.init;
+
+    T meanRes = 0;
+    size_t i = 0;
+
+    // Knuth & Welford mean calculation
+    // division per element is slower, but more accurate
+    foreach (e; r)
+    {
+        immutable T delta = (e - meanRes);
+        meanRes += delta / ++i;
+    }
+
+    return meanRes;
+}
+
+/// ditto
+auto mean(R, T)(R r, T seed)
+    if (isInputRange!R &&
+        !isNumeric!(ElementType!R) &&
+        is(typeof(r.front + seed)) &&
+        is(typeof(r.front / size_t(1))) &&
+        !isInfinite!R)
+{
+    import std.algorithm.iteration : reduce;
+    import std.typecons : tuple;
+
+    // per item division vis-a-vis the previous overload is too
+    // inaccurate for integer division, which the user defined
+    // types might be representing
+    auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+        (tuple(size_t(0), seed), r);
```

Comment:
Hmm AFAICT this only makes sense for `BigInt` (or similar types with arbitrary precision), for all other user-defined or subtyped types using `real` and the first overload makes more sense...

Also there's no need to save and compute the `length` is `hasLength` is available. Another lambda comes with a cost.